Sources/Datadog/DatadogCore/Context/DatadogContextProvider.swift

```diff
@@ -0,0 +1,95 @@
+
+    /// The queue used to synchronize the access to the `DatadogContext`.
+    ///
+    /// Concurrent queue is used for performant reads, `.barrier` must be used
+    /// to make writes exclusive.
+    private let queue = DispatchQueue(
+        label: "com.datadoghq.core-context",
+        attributes: .concurrent
+    )
+
+    private var subscriptions: [ContextValueSubscription] = []
+    private var networkConnectionInfoReader: AnyContextValueReader<NetworkConnectionInfo?>
+    private var carrierInfoReader: AnyContextValueReader<CarrierInfo?>
+
+    /// Creates a context provider to perform reads and writes on the
+    /// shared Datadog context.
+    ///
+    /// - Parameters:
+    ///   - context:                        The context inital context value.
+    ///   - serverOffsetPublisher:          The server date publisher to synchronize the `context.serverTimeOffset`
+    ///                                     parameter.
+    ///   - networkConnectionInfoPublisher: The network info publisher to synchronize the
+    ///                                     `context.networkConnectionInfo` parameter.
+    ///   - carrierInfoPublisher:           The carrier info publisher to synchronize the `context.carrierInfo`
+    ///                                     parameter.
+    init(
+        context: DatadogContext,
+        serverOffsetPublisher: ServerOffsetPublisher,
+        networkConnectionInfoPublisher: AnyNetworkConnectionInfoPublisher,
+        carrierInfoPublisher: AnyCarrierInfoPublisher
+    ) {
+        self.context = context
+        self.context.serverTimeOffset = serverOffsetPublisher.initialValue
+        self.networkConnectionInfoReader = networkConnectionInfoPublisher.eraseToAnyReader()
+        self.carrierInfoReader = carrierInfoPublisher.eraseToAnyReader()
+
+        self.subscriptions = [
+            serverOffsetPublisher.subscribe { offset in
+                self.write { $0.serverTimeOffset = offset }
+            },
+            networkConnectionInfoPublisher.set(queue: queue).subscribe { info in
+                self.write { $0.networkConnectionInfo = info }
+            },
+            carrierInfoPublisher.subscribe { info in
+                self.write { $0.carrierInfo = info }
+            },
+        ]
+    }
+
+    deinit {
+        subscriptions.forEach { $0.cancel() }
+    }
+
+    /// Reads to the `context` asynchronously, without blocking the caller thread.
+    func read(block: @escaping (DatadogContext) -> Void) {
+        queue.async {
+            var context = self.context
+            self.networkConnectionInfoReader.read { context.networkConnectionInfo = $0 }
+            self.carrierInfoReader.read { context.carrierInfo = $0 }
+            block(context)
+        }
+    }
+
+    /// Writes to the `context` asynchronously, without blocking the caller thread.
+    func write(block: @escaping (inout DatadogContext) -> Void) {
+        queue.async(flags: .barrier) { block(&self.context) }
+    }
```

Comment:
I cannot foresee the usage yet 🤔💭. Will `read(block:)` be used from _"get event write context"_ call? Will `write(block:)` be used when Features inject their portions of _SDK context_?

Reply:
Then I wonder what will be pattern of assembling context when we introduce other publishers (_user info_, _tracking consent_, _launch time_, _app state_, ...)?

In general, we have **two kinds** of publishers (in V1 called "providers"):
* **passive** ones, that require **pulling** the value (like _carrier info provider_ for iOS 11);
* **active** ones, that **publish** an update to the value (like _network info provider_ for iOS 12 or _user info provider_).

Does this pattern in `read(block:)`:
```swift
self.networkConnectionInfoReader.read { context.networkConnectionInfo = $0 }
self.carrierInfoReader.read { context.carrierInfo = $0 }
```
express that both _network_ and _carrier info_ publishers are passive, so we need to pull their value? Is this because the kind of these publishers depends on their implementation (iOS 11 / 12+)?

When it comes to active publishers, we won't need to call them in `read(block:)`, because they will update the `context` from `write(block:)` (managed through subscription). Is this correct 🤔💭?

Reply:
Yes, that is correct!
As we discussed, I will split reader from publisher for _network_ and _carrier info_, and other that require both behaviours. It will make it clear from `DatadogContextProvider` what is active and what is passive 👍 

I will also update the `ContextValueReader` interface so it differs more from the `ContextValuePublisher`.